libsignal-service/src/push_service.rs

```diff
@@ -72,7 +72,7 @@ pub const STICKER_PATH: &str = "stickers/%s/full/%d";
 pub const KEEPALIVE_TIMEOUT_SECONDS: Duration = Duration::from_secs(55);
 pub const DEFAULT_DEVICE_ID: u32 = 1;
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize)]
```

Comment:
It's not necessarily immediately clear to me how this enum gets serialized; shouldn't we just be using `to_string()` on instances?

Reply:
I'll have to double check why that was necessary. I added those when I tried to make this more Diesel-friendly, but `Display` also exists currently.

Reply:
It would be this (cleaned up):

```
   Compiling libsignal-service v0.1.0 (/home/matti/SFOS/libsignal-service-rs/libsignal-service)
error[E0277]: the trait bound `ServiceIdType: Serialize` is not satisfied
    --> /home/matti/SFOS/libsignal-service-rs/libsignal-service/src/service_address.rs:18:45
     |
18   | #[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
     |                                             ^^^^^^^^^ the trait `Serialize` is not implemented for `ServiceIdType`
...
21   |     pub identity: ServiceIdType,
     |     --- required by a bound introduced by this call
     |
note: required by a bound in `account_manager::_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/matti/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.197/src/ser/mod.rs:1865:12
     |
1859 |     fn serialize_field<T: ?Sized>(
     |        --------------- required by a bound in this associated function
...
1865 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`

error[E0277]: the trait bound `ServiceIdType: Deserialize<'_>` is not satisfied
    --> /home/matti/SFOS/libsignal-service-rs/libsignal-service/src/service_address.rs:21:19
     |
21   |     pub identity: ServiceIdType,
     |                   ^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `ServiceIdType`
     |
note: required by a bound in `next_element`
    --> /home/matti/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.197/src/de/mod.rs:1726:12
     |
1724 |     fn next_element<T>(&mut self) -> Result<Option<T>, Self::Error>
     |        ------------ required by a bound in this associated function
1725 |     where
1726 |         T: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `SeqAccess::next_element`

error[E0277]: the trait bound `ServiceIdType: Deserialize<'_>` is not satisfied
    --> /home/matti/SFOS/libsignal-service-rs/libsignal-service/src/service_address.rs:21:19
     |
21   |     pub identity: ServiceIdType,
     |                   ^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `ServiceIdType`
     |
note: required by a bound in `next_value`
    --> /home/matti/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.197/src/de/mod.rs:1865:12
     |
1863 |     fn next_value<V>(&mut self) -> Result<V, Self::Error>
     |        ---------- required by a bound in this associated function
1864 |     where
1865 |         V: Deserialize<'de>,
     |            ^^^^^^^^^^^^^^^^ required by this bound in `MapAccess::next_value`

error[E0277]: the trait bound `ServiceIdType: Deserialize<'_>` is not satisfied
  --> /home/matti/SFOS/libsignal-service-rs/libsignal-service/src/service_address.rs:21:5
   |
21 |     pub identity: ServiceIdType,
   |     ^^^ the trait `Deserialize<'_>` is not implemented for `ServiceIdType`
   |
note: required by a bound in `account_manager::_::_serde::__private::de::missing_field`
  --> /home/matti/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.197/src/private/de.rs:25:8
   |
23 | pub fn missing_field<'de, V, E>(field: &'static str) -> Result<V, E>
   |        ------------- required by a bound in this function
24 | where
25 |     V: Deserialize<'de>,
   |        ^^^^^^^^^^^^^^^^ required by this bound in `missing_field`
```
So, I guess since it goes down to Diesel, (`SeqAccess::next_element`) they have to be kept in?

Reply:
I don't think Diesel uses Serialize. I did some digging.

`ServiceAddress` now has the `ServiceIdType` as field. `ServiceAddress` gets serialized when calling `PushService::send_messages`. However, the Signal Server *ignores* unknown fields, and it never reads the `pub recipient: ServiceAddress,` which we currently have, nor the `@JsonProperty private String destination;` which Signal Android currently uses as a replacement.

So to mimic the Android client better, I would suggest:

```diff
diff --git a/libsignal-service/src/push_service.rs b/libsignal-service/src/push_service.rs
index 9cd10d12a..a2383c843 100644
--- a/libsignal-service/src/push_service.rs
+++ b/libsignal-service/src/push_service.rs
@@ -75,7 +75,7 @@ pub const STICKER_PATH: &str = "stickers/%s/full/%d";
 pub const KEEPALIVE_TIMEOUT_SECONDS: Duration = Duration::from_secs(55);
 pub const DEFAULT_DEVICE_ID: u32 = 1;
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Serialize, Deserialize, Hash)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
 pub enum ServiceIdType {
     /// Account Identity (ACI)
     ///
@@ -811,7 +811,7 @@ pub trait PushService: MaybeSend {
         &mut self,
         messages: OutgoingPushMessages,
     ) -> Result<SendMessageResponse, ServiceError> {
-        let path = format!("/v1/messages/{}", messages.recipient.uuid);
+        let path = format!("/v1/messages/{}", messages.destination);
         self.put_json(
             Endpoint::Service,
             &path,
diff --git a/libsignal-service/src/sender.rs b/libsignal-service/src/sender.rs
index 47c408722..c2218c6fa 100644
--- a/libsignal-service/src/sender.rs
+++ b/libsignal-service/src/sender.rs
@@ -37,7 +37,7 @@ pub struct OutgoingPushMessage {
 
 #[derive(serde::Serialize, Debug)]
 pub struct OutgoingPushMessages {
-    pub recipient: ServiceAddress,
+    pub destination: uuid::Uuid,
     pub timestamp: u64,
     pub messages: Vec<OutgoingPushMessage>,
     pub online: bool,
@@ -498,7 +498,7 @@ where
                 .await?;
 
             let messages = OutgoingPushMessages {
-                recipient,
+                destination: recipient.uuid,
                 timestamp,
                 messages,
                 online,
diff --git a/libsignal-service/src/service_address.rs b/libsignal-service/src/service_address.rs
index caa22364d..4da35634e 100644
--- a/libsignal-service/src/service_address.rs
+++ b/libsignal-service/src/service_address.rs
@@ -1,7 +1,6 @@
 use std::convert::TryFrom;
 
 use libsignal_protocol::{DeviceId, ProtocolAddress};
-use serde::{Deserialize, Serialize};
 use uuid::Uuid;
 
 pub use crate::push_service::ServiceIdType;
@@ -15,7 +14,7 @@ pub enum ParseServiceAddressError {
     NoUuid,
 }
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
 pub struct ServiceAddress {
     pub uuid: Uuid,
     pub identity: ServiceIdType,
diff --git a/libsignal-service/src/websocket/sender.rs b/libsignal-service/src/websocket/sender.rs
index 166d9c800..e9ea90ef6 100644
--- a/libsignal-service/src/websocket/sender.rs
+++ b/libsignal-service/src/websocket/sender.rs
@@ -12,7 +12,7 @@ impl SignalWebSocket {
         &mut self,
         messages: OutgoingPushMessages,
     ) -> Result<SendMessageResponse, ServiceError> {
-        let path = format!("/v1/messages/{}", messages.recipient.uuid);
+        let path = format!("/v1/messages/{}", messages.destination);
         self.put_json(&path, messages).await
     }
 
@@ -21,7 +21,7 @@ impl SignalWebSocket {
         messages: OutgoingPushMessages,
         access: &UnidentifiedAccess,
     ) -> Result<SendMessageResponse, ServiceError> {
-        let path = format!("/v1/messages/{}", messages.recipient.uuid);
+        let path = format!("/v1/messages/{}", messages.destination);
         let header = format!(
             "Unidentified-Access-Key:{}",
             BASE64_RELAXED.encode(&access.key)

```

... although we could actually get rid of the `destination` field altogether (but so should Android!).

Reply:
Maybe the `uuid::Uuid` field should also be a `String`, such that we can have the `PNI:` prefix in there? I'm not sure how that works though, so that's your call. I don't think it'll matter in this case, because I expect the behaviour to be the same!

Reply:
Your suggestion seems to work and it doesn't change anything at this point. Let's change it to `String` if and when the time comes.